modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.index.mapper.TypeParsers.nodeIndexOptionValue;
+import static org.elasticsearch.index.mapper.TypeParsers.parseMultiField;
+
+public class SearchAsYouTypeFieldMapper extends FieldMapper implements ArrayValueMapperParser {
+
+    private static final Logger LOG = LogManager.getLogger(SearchAsYouTypeFieldMapper.class);
+
+    public static class Defaults {
+
+        public static final int MIN_GRAM = 1;
+        public static final int MAX_GRAM = 20;
+        public static final int MAX_SHINGLE_SIZE = 3;
+
+        public static final MappedFieldType FIELD_TYPE = new SearchAsYouTypeFieldType();
+
+        static {
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<?, ?> parse(String name,
+                                          Map<String, Object> node,
+                                          ParserContext parserContext) throws MapperParsingException {
+
+            final SearchAsYouTypeFieldMapper.Builder builder = new SearchAsYouTypeFieldMapper.Builder(name);
+
+            NamedAnalyzer analyzer = parserContext.getIndexAnalyzers().getDefaultIndexAnalyzer();
+            int maxShingleSize = Defaults.MAX_SHINGLE_SIZE;
+
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                final Map.Entry<String, Object> entry = iterator.next();
+                final String fieldName = entry.getKey();
+                final Object fieldNode = entry.getValue();
+                if (fieldName.equals("type")) {
+                    continue;
+                } else if (fieldName.equals("index_options")) {
+                    builder.indexOptions(nodeIndexOptionValue(fieldNode));
+                    iterator.remove();
+                } else if (fieldName.equals("fields")) {
+                    parseMultiField(builder, name, parserContext, fieldName, fieldNode);
+                    iterator.remove();
+                } else if (fieldName.equals("analyzer")) {
+                    final String analyzerName = fieldNode.toString();
+                    analyzer = parserContext.getIndexAnalyzers().get(analyzerName);
+                    if (analyzer == null) {
+                        throw new MapperParsingException("Analyzer [" + analyzerName + "] not found for field  [" + name + "]");
+                    }
+                    iterator.remove();
+                } else if (fieldName.equals("max_shingle_size")) {
+                    maxShingleSize = XContentMapValues.nodeIntegerValue(fieldNode);
+                    iterator.remove();
+                }
+            }
```

Comment:
Should we throw an error if the map is not empty (has some fields that we don't support) ?

Reply:
I assumed whatever calls this method did that afterwards, it looks like in at least some places they do. Should I throw one here too? https://github.com/elastic/elasticsearch/blob/3c03975ce396ff3e6f27b79d26e77a4cf1081f67/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L276-L284